pennylane_lightning/core/src/utils/NDPermuter.hpp

```diff
@@ -0,0 +1,268 @@
+            idx_old == idx_new,
+            "New indices are an invalid permutation of the existing indices");
+
+        permuter_b_.Transpose(data_in, shape, data_out, current_order,
+                              new_order);
+    }
+
+    /**
+     * @brief Reshape the given lexicographic data vector from old to new index
+     * ordering.
+     *
+     * @tparam T Data participating in the permutation.
+     * @param data_in Input data to be transposed.
+     * @param shape Current shape of the tensor data in each dimension.
+     * @param current_order Current index ordering of the tensor.
+     * @param new_order New index ordering of the tensor.
+     * @return std::vector<T> Output data following the transpose.
+     */
+    template <class T>
+    std::vector<T> Transpose(const std::vector<T> &data_in,
+                             const std::vector<size_t> &shape,
+                             const std::vector<std::string> &current_order,
+                             const std::vector<std::string> &new_order) {
+        const std::set<std::string> idx_old(current_order.begin(),
+                                            current_order.end());
+        const std::set<std::string> idx_new(new_order.begin(), new_order.end());
+        const auto data_size = std::accumulate(shape.begin(), shape.end(), 1,
+                                               std::multiplies<std::size_t>());
+        PL_ABORT_IF_NOT(idx_old.size() == current_order.size(),
+                        "Duplicate existing indices found. Please ensure "
+                        "indices are unique.");
+        PL_ABORT_IF_NOT(idx_new.size() == new_order.size(),
+                        "Duplicate transpose indices found. Please ensure "
+                        "indices are unique.");
+        PL_ABORT_IF_NOT(shape.size() == new_order.size(),
+                        "Tensor shape does not match number of indices.");
+        PL_ABORT_IF_NOT(data_size == data_in.size(),
+                        "Tensor shape does not match given tensor data.");
+        PL_ABORT_IF_NOT(
+            idx_old == idx_new,
+            "New indices are an invalid permutation of the existing indices");
+
+        PL_ABORT_IF(shape.empty(), "Tensor shape cannot be empty.");
+        PL_ABORT_IF(new_order.empty(), "Tensor indices cannot be empty.");
+        return permuter_b_.Transpose(data_in, shape, current_order, new_order);
+    }
+
+  protected:
+    friend PermuterBackend;
+
+  private:
+    PermuterBackend permuter_b_;
+};
+
+/**
+ * @brief Default Permuter backend class for generalised transforms. Adapted
+ * from QFlex.
+ *
+ * @tparam blocksize Controls the internal data chunk size for cache blocking.
+ */
+template <size_t BLOCKSIZE = 1024> class DefaultPermuter {
+
+  public:
+    /**
+     * @brief Reference-based transpose operation. See `Permuter` class for more
+     * details.
+     */
+    template <class T>
+    void Transpose(const std::vector<T> &data_,
```

Comment:
did you mean `data_in`?

```suggestion
    void Transpose(const std::vector<T> &data_in,
```

Reply:
To be resolved by later importing directly from JET. Resolving.